pkg/fqdn/service/service.go

```diff
@@ -0,0 +1,446 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package service
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/netip"
+
+	"github.com/cilium/dns"
+	"github.com/google/uuid"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/keepalive"
+
+	"github.com/cilium/cilium/pkg/clustermesh/types"
+	"github.com/cilium/cilium/pkg/container/versioned"
+	"github.com/cilium/cilium/pkg/endpoint"
+	"github.com/cilium/cilium/pkg/endpointmanager"
+	"github.com/cilium/cilium/pkg/fqdn/dnsproxy"
+	"github.com/cilium/cilium/pkg/identity"
+	"github.com/cilium/cilium/pkg/ipcache"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/policy"
+	"github.com/cilium/cilium/pkg/time"
+
+	standalonednsproxy "github.com/cilium/cilium/api/v1/standalone-dns-proxy"
+)
+
+type updateOnDNSMsgFunc func(lookupTime time.Time, ep *endpoint.Endpoint, qname string, responseIPs []netip.Addr, TTL int, stat *dnsproxy.ProxyRequestContext) error
+
+type FQDNDataServer struct {
+	standalonednsproxy.UnimplementedFQDNDataServer
+
+	ctx             context.Context
+	closeServer     func()
+	endpointManager endpointmanager.EndpointManager
+
+	// streams is a map of the active streams and their cancel functions
+	// Streams are added when a client(standalone dns proxy) subscribes to DNS policies and removed when the client closes the connection
+	streams lock.Map[standalonednsproxy.FQDNData_StreamPolicyStateServer, context.CancelFunc]
```

Comment:
Keeping this state server-scoped seems unnecessary - the cancel function (like the server) is request-scoped, so we can just pass it to child functions of the request handler?

Reply:
If we don't keep the state, then we have to pass it to every request handler. For closing the server, we need to cancel each stream. I would want to keep the state as it if that;s okay ?